smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/DefaultEndpointProviderGenerator.kt

```diff
@@ -0,0 +1,319 @@
+            writer.withBlock("return #T(", ")", RuntimeTypes.Http.Endpoints.Endpoint) {
+                writeInline("#T.parse(", RuntimeTypes.Http.Url)
+                writeExpression(rule.endpoint.url)
+                write("),")
+
+                if (rule.endpoint.headers.isNotEmpty()) {
+                    withBlock("headers = #T {", "},", RuntimeTypes.Http.Headers) {
+                        rule.endpoint.headers.entries.forEach { (k, v) ->
+                            v.forEach {
+                                writeInline("append(#S, ", k)
+                                writeExpression(it)
+                                write(")")
+                            }
+                        }
+                    }
+                }
+
+                if (rule.endpoint.properties.isNotEmpty()) {
+                    withBlock("attributes = #T().apply {", "},", RuntimeTypes.Utils.Attributes) {
+                        rule.endpoint.properties.entries.forEach { (k, v) ->
+                            writeInline("set(#T(#S), ", RuntimeTypes.Utils.AttributeKey, k.asString())
+                            writeExpression(v)
+                            write(")")
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun renderErrorRule(rule: ErrorRule) {
+        withConditions(rule.conditions) {
+            writer.writeInline("throw #T(", RuntimeTypes.Http.Endpoints.EndpointProviderException)
+            writeExpression(rule.error)
+            writer.write(")")
+        }
+    }
+
+    private fun renderTreeRule(rule: TreeRule) {
+        withConditions(rule.conditions) {
+            rule.rules.forEach(::renderRule)
+        }
+    }
+
+    override fun visitInteger(value: Int) {
+        writer.writeInline("#L", value)
+    }
+
+    override fun visitString(value: Template) {
+        writer.writeInline("\"")
+        value.accept(this).toList() // must "consume" the stream to actually generate everything
+        writer.writeInline("\"")
+    }
+
+    override fun visitBool(value: Boolean) {
+        writer.writeInline("#L", value)
+    }
+
+    override fun visitRecord(value: MutableMap<Identifier, Literal>) {
+        writer.withBlock("#T().apply {", "}", RuntimeTypes.Utils.Attributes) {
+            value.entries.forEach { (k, v) ->
+                writeInline("set(#T(#S), ", RuntimeTypes.Utils.AttributeKey, k.asString())
+                v.accept(this@DefaultEndpointProviderGenerator)
+                write(")")
+            }
+        }
+    }
+
+    override fun visitTuple(value: MutableList<Literal>) {
+        if (value.size == 1) { // unclear why this is a thing
+            value[0].accept(this)
+            return
+        }
+
+        writer.withBlock("listOf(", ")") {
+            value.forEach {
+                it.accept(this@DefaultEndpointProviderGenerator)
+                write(",")
+            }
+        }
+    }
```

Comment:
**Question**: Why should a 1-tuple get rendered as a scalar but all other tuples would get rendered as a list? Is this part of the EP2.0 spec?